### PR TITLE
nixos/switch-to-configuration: add sysinit-reactivation.target

### PR DESCRIPTION
**Pull request by nikstur**
## Description of changes

Second try of https://github.com/NixOS/nixpkgs/pull/269983

This solution is my interpretation of what @ElvishJerricco described in this comment: https://github.com/NixOS/nixpkgs/pull/269983#issuecomment-1832592460

Although this doesnt remove any custom code from stc, now at least we can restart multiple services that are pulled in by sysinit.target while respecting the ordering between them (e.g. sysusers runs before tmpfiles).

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by nikstur** [comment]
@ofborg test switchTest acme

File: nixos/modules/system/activation/switch-to-configuration.pl [hunk 1]
```diff
@@ -889,9 +889,16 @@ sub filter_units {
...
+# Restart sysinit-reactivation.target.
+# This target only exists to restart services ordered before sysinit.target. We
+# cannot use X-StopOnReconfiguration to restart sysinit.target because then ALL
+# services of the system would be restarted since all normal services have a
+# default dependency on sysinit.target. This target ensures that services
+# ordered before sysinit.target get started in the correct order, most notably
+# BEFORE the rest of the services. Ordering between these services is
+# respected.
+print STDERR "restarting sysinit-reactivation.target\n";
+system("$new_systemd/bin/systemctl", "restart", "sysinit-reactivation.target") == 0 or $res = 4;
```
Comment (Mic92) [inline nixos/modules/system/activation/switch-to-configuration.pl:901]: Can we also use this than for agenix and sops-nix?

Comment (Mic92) [reply nixos/modules/system/activation/switch-to-configuration.pl:901]: This gives us better visibility when monitoring a service I would say. One can export service state as metrics and also export logs.

Comment (nikstur) [reply nixos/modules/system/activation/switch-to-configuration.pl:901]: No the big argument is that we can now respect the ordering between the services pulled in by sysinit. Concretely, this allows us to restart sysusers BEFORE tmpfiles which wasn't previously possible.

**Comment by nikstur** [comment]
I'll add a specific test case.

**Comment by ElvishJerricco** [comment]
@dasJ Although this doesn't *directly* contribute to removing perl from activation, I think it's an important step. Having a logic for reconfiguring the system via systemd seems important to me.

File: nixos/doc/manual/development/what-happens-during-a-system-switch.chapter.md [hunk 2]
```diff
@@ -37,7 +37,7 @@ of actions is always the same:
 - Forget about the failed state of units (`systemctl reset-failed`)
 - Reload systemd (`systemctl daemon-reload`)
 - Reload systemd user instances (`systemctl --user daemon-reload`)
-- Set up tmpfiles (`systemd-tmpfiles --create`)
+- Reactivate sysinit (systemctl restart sysinit-reactivation.target)
```
Comment (cole-h) [inline nixos/doc/manual/development/what-happens-during-a-system-switch.chapter.md:40]: nit: (the rest of these parentheticals are backquoted, might as well stay consistent)

```suggestion
- Reactivate sysinit (`systemctl restart sysinit-reactivation.target`)
```

**Comment by AleXoundOS** [comment]
Does this PR support the case when systemd-tmpfiles creates files in a filesystem location, which must be mounted first?
It looks like switch-to-configuration.pl works in the following order:
1. [parses fstab](https://github.com/NixOS/nixpkgs/blob/c4e2e52a2113b59cf58da441f6833fd86c682f71/nixos/modules/system/activation/switch-to-configuration.pl#L659), but does not mount filesystems (leaving this to `local-fs.target`)
2. [runs systemd-tmpfiles command](https://github.com/NixOS/nixpkgs/blob/c4e2e52a2113b59cf58da441f6833fd86c682f71/nixos/modules/system/activation/switch-to-configuration.pl#L894)
3. [restarts](https://github.com/NixOS/nixpkgs/blob/c4e2e52a2113b59cf58da441f6833fd86c682f71/nixos/modules/system/activation/switch-to-configuration.pl#L924C43-L924C43) and starts systemd units, perhaps one of which is `local-fs.target`?

I'm not sure about `local-fs.target`, but it looks like `systemd-tmpfiles` is called before mounting...

**Comment by nikstur** [comment]
> It's probably not a huge issue in the end, but maybe we should at least mention this in the release notes?

@r-vdp I added a release note. Hope that describes the issue clearly enough.

> Does this PR support the case when systemd-tmpfiles creates files in a filesystem location, which must be mounted first?

@AleXoundOS If I understand this correctly, this is indeed a limitation of the previous system because the upstream systemd-tmpfiles unit is configured to run after local-fs.target but this is not respected right now. With the changes I propose, you can indeed order this correctly but you'd need to add the dependency to local-fs.target as described in the docs. I'd be happy to review a PR if you're interested in doing this.

**Comment by nikstur** [comment]
@ofborg test sysinit-reactivation acme switchTest

**Comment by nikstur** [comment]
@dasJ can I get you to review this one again? I added a test as you requested.

**Comment by AleXoundOS** [comment]
> With the changes I propose, you can indeed order this correctly but you'd need to add the dependency to local-fs.target as described in the docs.

I'm afraid, that simply making systemd-tmpfiles dependent on FS mounts will potentially introduce locking of OS booting, e.g. if some non-crucial external FS is unavailable. It seems running systemd-tmpfiles is mandatory for OS internals, even systemd itself to work properly.
Maybe in order to circumvent locking of OS booting entirely we need to use `--exclude-prefix=PATH` option to ignore non-system rules and use `--prefix=PATH` for additional runs of `systemd-tmpfiles`. Thus, being more granular in dependencies. But I'm not sure whether it is the right approach.

**Comment by nikstur** [comment]
> I'm afraid, that simply making systemd-tmpfiles dependent on FS mounts will potentially introduce locking of OS booting,

systemd-tmpfiles [already depends on local-fs.target](https://github.com/systemd/systemd/blob/main/units/systemd-tmpfiles-setup.service#L15)

**Comment by r-vdp** [comment]
@nikstur I think there's an issue still because `systemd-setup-tmpfiles.service` calls `systemd-tmpfiles` with the `--boot` option, but this can break running systems when invoked during a switch because it also cleans up directories that were explicitly marked to be ignored for running systems.

For instance, dbus-broker fails when reloaded during a switch when I run with this patch, because its temporary files get wiped, presumably because we have lines like
```
R! /tmp/systemd-private-*
R! /var/tmp/systemd-private-*
```
in our tmpfiles config, which are supposed to be executed only during boot, but with this patch they are also executed during a switch.

I think we need a distinction between the tmpfiles invocation during bootup, which has `--boot`, and the one that we do while the system is running (like during a switch), which doesn't set `--boot`.

**Comment by nikstur** [comment]
I now added a separate service `systemd-tmpfiles-resetup.service` to encode restarting tmpfiles without `--boot`. Hope this fixes it. 

>  dbus-broker fails is there a test you can point me to? I can't think of another way of verifying this.

**Comment by r-vdp** [comment]
> I now added a separate service `systemd-tmpfiles-resetup.service` to encode restarting tmpfiles without `--boot`. Hope this fixes it. 
> 
> >  dbus-broker fails is there a test you can point me to? I can't think of another way of verifying this.

I am running with the updated patch now on my system to test.

I can try to reproduce this in a VM test, I think it should be sufficient to restart the tmpfiles service (the one with `--boot`) and then to reload dbus-broker.

**Comment by r-vdp** [comment]
@nikstur: [this commit](https://github.com/r-vdp/nixpkgs/commit/16a7b0373860ab9c73b8edeff2b43069580ab7e1) introduces a test that fails with the previous implementation and succeeds now, showing the problem with dbus-broker.

I'm sure that there must be other services that would break if their temporary files get wiped during a switch, dbus-broker is just the only one that I encountered the issue with on my own system.

**Comment by nikstur** [comment]
> @nikstur: [this commit](https://github.com/r-vdp/nixpkgs/commit/16a7b0373860ab9c73b8edeff2b43069580ab7e1) introduces a test that fails with the previous implementation and succeeds now, showing the problem with dbus-broker.

Nice! Thank you. 

So this PR is good to go from your side @r-vdp?

**Comment by r-vdp** [comment]
Yeah I think so, the change to not include `--boot` fixes the one issue that I encountered and I've been running my main system with this patch for several weeks now.

Maybe good to cherry-pick the nixos test onto this branch to avoid regressions.

**Review by r-vdp — APPROVED** [review]

File: nixos/doc/manual/release-notes/rl-2405.section.md [hunk 3]
```diff
@@ -51,6 +51,17 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
...
+- switch-to-configuration does not directly call systemd-tmpfiles anymore.
+  Instead the new artificial sysinit-reactivation.target is introduced which
+  allows to restart multiple services that are ordered before sysinit.target
+  and respect the ordering between the services.
+  `systemd-tmpfiles-setup.service` now sets `RefuseManualStop = false` so that
+  the unit can be restarted by switch-to-configuration. When upgrading via a
+  system switch, however, this will lead to a failure because on the previous
+  generation it is still `RefuseManualStop = true`. See [this
+  issue](https://github.com/NixOS/nixpkgs/pull/271067#issuecomment-1845714645)
+  for context. A reboot will solve this issue.
```
Comment (ElvishJerricco) [inline nixos/doc/manual/release-notes/rl-2405.section.md:63]: This bit about `RefuseManualStop` is not actually true anymore in the current state of this PR.

File: nixos/modules/system/boot/systemd/tmpfiles.nix [hunk 4]
```diff
@@ -150,6 +150,39 @@ in
...
+    # Allow systemd-tmpfiles to be restarted by switch-to-configuration. This
+    # service is not pulled into the normal boot process. It only exists for
+    # switch-to-configuration.
+    #
+    # This needs to be a separate unit because it does not execute
+    # systemd-tmpfiles with `--boot` as that is supposed to only be executed
+    # once at boot time.
+    #
+    # Keep this aligned with the upstream `systemd-tmpfiles-setup.service` unit.
+    systemd.services."systemd-tmpfiles-resetup" = {
```
Comment (ElvishJerricco) [inline nixos/modules/system/boot/systemd/tmpfiles.nix:162]: Needs a `description`.

**Review by ElvishJerricco — DISMISSED** [review]
A little annoyance here is that reactivation units are being start twice on stc. Once from `systemctl restart sysinit-reactivation.target` and once from `starting units ....`. I don't know how much this matter realistically, just a little imperfection IMO. I don't mind ignoring it.

**Comment by Ramblurr** [comment]
For those of us that are still on nixos stable (23.11) and therefore do not have `systemd-tmpfiles-resetup.service` and `sysinit-reactivation.target`. Is there anyway to ensure that my own systemd service is run before the tmpfiles rules are updated during a normal rebuild switch (i.e.,  not just at boot time)?

**Pull request merged**